app/models/form.rb

```diff
@@ -1,171 +1,51 @@
 class Form
-  attr_reader :form_definition
+  attr_reader :form_definition, :sections, :subsections, :pages, :questions
 
   def initialize(form_path)
     raise "No form definition file exists for given year".freeze unless File.exist?(form_path)
 
     @form_definition = JSON.parse(File.open(form_path).read)
+    @sections = form_definition["sections"].map { |id, s| Form::Section.new(id, s, self) }
+    @subsections = sections.flat_map(&:subsections)
+    @pages = subsections.flat_map(&:pages)
+    @questions = pages.flat_map(&:questions)
   end
 
-  # Returns a hash with sections as keys
-  def all_sections
-    @all_sections ||= @form_definition["sections"]
+  def get_subsection(id)
+    subsections.find { |s| s.id == id }
   end
 
-  # Returns a hash with subsections as keys
-  def all_subsections
-    @all_subsections ||= all_sections.map { |_section_key, section_value|
-      section_value["subsections"]
-    }.reduce(:merge)
-  end
-
-  # Returns a hash with pages as keys
-  def all_pages
-    @all_pages ||= all_subsections.map { |_subsection_key, subsection_value|
-      subsection_value["pages"]
-    }.reduce(:merge)
-  end
-
-  # Returns a hash with the pages of a subsection as keys
-  def pages_for_subsection(subsection)
-    all_subsections[subsection]["pages"]
-  end
-
-  # Returns a hash with the questions as keys
-  def questions_for_page(page)
-    all_pages[page]["questions"]
-  end
-
-  # Returns a hash with the questions as keys
-  def questions_for_subsection(subsection)
-    pages_for_subsection(subsection).map { |title, _value| questions_for_page(title) }.reduce(:merge)
-  end
-
-  # Returns a hash with soft validation questions as keys
-  def soft_validations_for_page(page)
-    all_pages[page]["soft_validations"]
-  end
-
-  def expected_responses_for_page(page)
-    questions_for_page(page).merge(soft_validations_for_page(page) || {})
-  end
-
-  def first_page_for_subsection(subsection)
-    pages_for_subsection(subsection).keys.first
+  def get_page(id)
+    pages.find { |p| p.id == id }
   end
 
   def subsection_for_page(page)
-    all_subsections.find { |_subsection_key, subsection_value|
-      subsection_value["pages"].key?(page)
-    }.first
-  end
-
-  def page_for_question(question)
-    all_pages.find { |_page_key, page_value| page_value["questions"].key?(question) }.first
+    subsections.find { |s| s.pages.find { |p| p.id == page.id } }
   end
 
   def next_page(page, case_log)
-    subsection = subsection_for_page(page)
-    page_idx = pages_for_subsection(subsection).keys.index(page)
-    nxt_page = pages_for_subsection(subsection).keys[page_idx + 1]
+    page_ids = subsection_for_page(page).pages.map(&:id)
+    page_idx = page_ids.index(page.id)
+    nxt_page = get_page(page_ids[page_idx + 1])
```

Comment:
I think these should be `page-index` and `next_page`

Reply:
Have changed `page_index`. `nxt_page` was to avoid shadowing the method name so left that one for now.